[c10d] switch ProcessGroup to be managed by intrusive_ptr

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42944 [c10d][jit] initial torchbind bindings for ProcessGroup
* **#47343 [c10d] switch ProcessGroup to be managed by intrusive_ptr**

Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

wanchaol committed:
[c10d] switch ProcessGroup to be managed by intrusive_ptr

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup to be managed by intrusive_ptr"


Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

[ghstack-poisoned]

agolynski commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -29,7 +29,7 @@ constexpr int kUnsetDivFactor = -1;
     std::vector<std::vector<torch::autograd::Variable>> replicas,
     std::vector<std::vector<size_t>> bucket_indices,
-    std::shared_ptr<c10d::ProcessGroup> process_group,
+    c10::intrusive_ptr<::c10d::ProcessGroup> process_group,
```
nit: keep c10d::ProcessGroup for consistency?

wanchaol committed:
Update on "[c10d] switch ProcessGroup to be managed by intrusive_ptr"


Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup to be managed by intrusive_ptr"


Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup to be managed by intrusive_ptr"


Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup to be managed by intrusive_ptr"


Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup to be managed by intrusive_ptr"


Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup to be managed by intrusive_ptr"


Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroup to be managed by intrusive_ptr"


Differential Revision: [D24723418](https://our.internmc.facebook.com/intern/diff/D24723418)

[ghstack-poisoned]
